File: src/workers/runner.ts
// Codex integration via the official @openai/codex-sdk threads API (SPEC §7).
//
// Workers are purely ephemeral: every run() is a FRESH Codex thread that runs one objective and is
// never resumed — there is no abort, no steer, no follow-up. We stream events so callers can record
// live progress, and report the thread id only as a trace artifact (eval/Inspector forensics).
//
// Auth: we never pass `apiKey`, so the SDK rides the cached ChatGPT-subscription login in
// CODEX_HOME (~/.codex/auth.json). Defense in depth: we strip OPENAI_API_KEY / CODEX_API_KEY
// from the env handed to the CLI so a stray key can't flip us to metered API billing.

import { spawn } from "node:child_process";
import { Codex, type ThreadItem } from "@openai/codex-sdk";
import type { Config } from "../config.js";
import { logger } from "../logger.js";

export interface CodexTurn {
  ok: boolean;
  /** The (single-use) thread's id — a trace artifact for forensics, never used to resume. */
  threadId?: string;
  /** Final agent message, suitable for sending back to the user. */
  finalResponse: string;
}

export type ProgressFn = (note: string) => void;

export interface CodexRunArgs {
  prompt: string;
  onProgress?: ProgressFn;
}

export interface CodexRunner {
  run(args: CodexRunArgs): Promise<CodexTurn>;
  loginStatus(): Promise<{ ok: boolean; detail: string }>;
}

const NO_OUTPUT = "(Codex produced no output.)";

/** Build the env handed to the Codex CLI: inherit everything except billing-flip keys, then layer
 *  on any extras (the manager passes LILA_MCP_TOKEN this way). Shared by the worker runner and the
 *  manager thread factory so both strip the keys identically. */
export function sanitizedEnv(extra: Record<string, string> = {}): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [k, v] of Object.entries(process.env)) {
    if (v !== undefined && k !== "OPENAI_API_KEY" && k !== "CODEX_API_KEY") env[k] = v;
  }
  return { ...env, ...extra };
}

/** Render a non-message thread item as a short live-progress line (or skip it). */
export function formatItem(item: ThreadItem): string | undefined {
  switch (item.type) {
    case "command_execution": {
      const cmd = item.command.replace(/\s+/g, " ").trim();
      return `$ ${truncate(cmd, 120)}`;
    }
    case "file_change": {
      const n = item.changes?.length ?? 0;
      return `✏️ ${n} file${n === 1 ? "" : "s"} changed`;
    }
    case "mcp_tool_call":
      return `🔧 ${item.server}.${item.tool}`;
    case "web_search":
      return `🔍 ${truncate(item.query, 120)}`;
    case "error":
      return `⚠️ ${truncate(item.message, 200)}`;
    default:
      // agent_message is the final answer (handled separately); reasoning/todo_list are noisy.
      return undefined;
  }
}

export function friendlyError(detail: string): string {
  const clipped = detail.slice(0, 1500);
  if (/usage limit|rate limit|quota|too many requests|429|purchase more credits/i.test(clipped)) {
    return (
      `⚠️ Codex hit a usage limit on the ChatGPT subscription. Wait for the window to reset or ` +
      `add credits/upgrade the plan backing the host's CODEX_HOME.\n\n${clipped}`
    );
  }
  const authish = /login|auth|401|unauthor|credential|expired/i.test(clipped);
  return authish
    ? `⚠️ Codex couldn't run — looks like an auth problem. Re-run \`codex login\` on the host ` +
        `(against the persistent CODEX_HOME).\n\n${clipped}`
    : `⚠️ Codex error.\n\n${clipped}`;
}

export function createCodexRunner(config: Config): CodexRunner {
  const codex = new Codex({
    env: sanitizedEnv(),
    ...(config.codexPathOverride ? { codexPathOverride: config.codexPathOverride } : {}),
  });

  const threadOptions = {
    workingDirectory: config.workspaceDir,
    skipGitRepoCheck: true,
    sandboxMode: config.sandboxMode,
    approvalPolicy: "never" as const,
  };

  return {
    async run({ prompt, onProgress }: CodexRunArgs): Promise<CodexTurn> {
      const thread = codex.startThread(threadOptions);

      logger.debug("Codex turn", { cwd: config.workspaceDir });

      // Hoisted out of the try so the catch below can prefer it: the Codex SDK throws on ANY non-zero
      // exit (including a clean turn.failed whose reason we already streamed), leaving only the
      // "Reading prompt from stdin..." banner on stderr — which would otherwise mask the real cause.
      let failure: string | undefined;
      try {
        const { events } = await thread.runStreamed(prompt);
        const agentMessages: string[] = [];

        for await (const event of events) {
          switch (event.type) {
            case "item.completed": {
              const item = event.item;
              if (item.type === "agent_message") agentMessages.push(item.text);
              else {
                const note = formatItem(item);
                if (note && onProgress) onProgress(note);
              }
              break;
            }
            case "turn.failed":
              failure = event.error?.message ?? "Codex turn failed.";
              break;
            case "error":
              failure = event.message;
              break;
            default:
              break;
          }
        }

        const threadId = thread.id ?? undefined;
        if (failure) return { ok: false, threadId, finalResponse: friendlyError(failure) };

        const finalResponse = agentMessages.join("\n\n").trim() || NO_OUTPUT;
        return { ok: true, threadId, finalResponse };
      } catch (err) {
        return {
          ok: false,
          threadId: thread.id ?? undefined,
          finalResponse: friendlyError(failure ?? (err as Error).message),
        };
      }
    },

    loginStatus: () => codexLoginStatus(config.codexPathOverride ?? "codex"),
  };
}

function truncate(s: string, max: number): string {
  return s.length > max ? `${s.slice(0, max - 1)}…` : s;
}

/**
 * Probe ChatGPT-subscription auth via `codex login status` (SPEC §4). The SDK has no
 * login-status API, so we shell out to the CLI (same CODEX_HOME, so the same auth).
 */
export function codexLoginStatus(
  bin: string,
  env: NodeJS.ProcessEnv = process.env,
): Promise<{ ok: boolean; detail: string }> {
  const childEnv = { ...env };
  delete childEnv.OPENAI_API_KEY;
  delete childEnv.CODEX_API_KEY;
  return new Promise((resolve) => {
    const child = spawn(bin, ["login", "status"], {
      env: childEnv,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let out = "";
    child.stdout.on("data", (d: Buffer) => (out += d.toString()));
    child.stderr.on("data", (d: Buffer) => (out += d.toString()));
    child.on("error", (err) => resolve({ ok: false, detail: err.message }));
    child.on("close", (code) => resolve({ ok: code === 0, detail: out.trim() }));
  });
}

File: src/workers/runner.rs
//! The worker runner seam: a single-shot backend run. Workers are purely ephemeral — every `run`
//! is a FRESH session that runs
//! one objective and is never resumed.

use std::path::PathBuf;

use async_trait::async_trait;
use thiserror::Error;

use crate::runtime::TokenUsage;

/// Inputs for a single worker run.
#[derive(Debug, Clone)]
pub struct RunArgs {
    /// The bare objective (standing rules come from the workspace `AGENTS.md`).
    pub prompt: String,
    /// Working directory for the run (the app workspace, or a subproject within it).
    pub cwd: PathBuf,
}

/// The result of a single worker run.
#[derive(Debug, Clone, Default)]
pub struct RunOutcome {
    pub ok: bool,
    /// Final agent message (the worker's summary, per the contract).
    pub final_response: String,
    /// The single-use session id — a trace artifact, never used to resume.
    pub thread_id: Option<String>,
    /// Token usage the worker's backend reported for the run (zero if the CLI reported none). Folded
    /// into the cumulative worker totals so prod + eval can see what the work tier actually cost.
    pub usage: TokenUsage,
}

/// Auth/login probe result.
#[derive(Debug, Clone)]
pub struct LoginStatus {
    pub ok: bool,
    pub detail: String,
}

/// Worker run errors (spawn/protocol/auth).
#[derive(Debug, Error)]
pub enum RunnerError {
    #[error("{0}")]
    Run(String),
}

/// Runs a single-shot worker objective.
#[async_trait]
pub trait Runner: Send + Sync {
    /// Run one objective to completion and return its summary.
    async fn run(&self, args: RunArgs) -> Result<RunOutcome, RunnerError>;
    /// Probe whether the backend is authenticated (subscription login present).
    async fn login_status(&self) -> LoginStatus;
}

/// Turn a non-zero-exit detail into owner-facing text (Codex wording).
pub fn friendly_error(detail: &str) -> String {
    let clipped: String = detail.chars().take(1500).collect();
    let lower = clipped.to_lowercase();
    if [
        "usage limit",
        "rate limit",
        "quota",
        "too many requests",
        "429",
        "purchase more credits",
    ]
    .iter()
    .any(|n| lower.contains(n))
    {
        return "⚠️ The agent hit a usage/rate limit. Wait for the quota to reset, or check \
                `codex login`."
            .to_string();
    }
    if lower.contains("login") || lower.contains("auth") || lower.contains("unauthorized") {
        return "⚠️ The agent isn't authenticated. Run `codex login` on the host.".to_string();
    }
    format!("⚠️ The agent run failed: {clipped}")
}

/// Claude-wording variant of [`friendly_error`].
pub fn friendly_claude_error(detail: &str) -> String {
    let clipped: String = detail.chars().take(1500).collect();
    let lower = clipped.to_lowercase();
    if [
        "usage limit",
        "rate limit",
        "quota",
        "too many requests",
        "429",
    ]
    .iter()
    .any(|n| lower.contains(n))
    {
        return "⚠️ Claude hit a usage/rate limit. Wait for the quota to reset, or check \
                `claude setup-token`."
            .to_string();
    }
    if lower.contains("login") || lower.contains("auth") || lower.contains("unauthorized") {
        return "⚠️ Claude isn't authenticated. Run `claude setup-token` on the host.".to_string();
    }
    format!("⚠️ The agent run failed: {clipped}")
}
